native/src/hooks/useLocation.ts

```diff
@@ -0,0 +1,141 @@
+      message: 'noPermission'
+    }
+  } else if (error.code === 2) {
+    return {
+      status: 'unavailable',
+      message: 'notAvailable'
+    }
+  } else {
+    return {
+      status: 'unavailable',
+      message: 'timeout'
+    }
+  }
+}
+
+type LocationInformationType = {
+  location: LocationType | null
+  locationState: LocationStateType
+  requestAndDetermineLocation: () => Promise<void>
+}
+
+const useLocation = (useSettingsListener = false): LocationInformationType => {
+  const [locationState, setLocationState] = useState<LocationStateType>({
+    status: 'unavailable',
+    message: 'loading'
+  })
+  const [location, setLocation] = useState<LocationType | null>(null)
+
+  const determineLocation = useCallback(() => {
+    Geolocation.getCurrentPosition(
+      (position: GeolocationResponse) => {
+        setLocation([position.coords.longitude, position.coords.latitude])
+        setLocationState({ status: 'ready' })
+      },
+      (error: GeolocationError) => {
+        setLocationState(locationStateOnError(error))
+      },
+      {
+        enableHighAccuracy: true,
+        timeout: 50000,
+        maximumAge: 3600000
+      }
+    )
+  }, [])
+
+  useEffect(() => {
+    checkLocationPermission().then(locationPermissionStatus => {
+      if (locationPermissionStatus === RESULTS.GRANTED) {
+        setLocationState({
+          message: 'loading',
+          status: 'unavailable'
+        })
+        determineLocation()
+      } else {
+        setLocationState({
+          status: 'unavailable',
+          message: 'noPermission'
+        })
+      }
+    })
+  }, [determineLocation])
+
+  const requestAndDetermineLocation = useCallback(async () => {
+    setLocationState({
+      message: 'loading',
+      status: 'unavailable'
+    })
+    const locationPermissionStatus = await checkLocationPermission()
+
+    if (locationPermissionStatus === RESULTS.BLOCKED) {
+      await openSettings()
+      setLocationState({
+        message: 'noPermission',
+        status: 'unavailable'
+      })
+    } else if (locationPermissionStatus === RESULTS.GRANTED) {
+      determineLocation()
+    } else {
+      if ((await requestLocationPermission()) === RESULTS.GRANTED) {
+        determineLocation()
+      } else {
+        setLocationState({
+          message: 'noPermission',
+          status: 'unavailable'
+        })
+      }
+    }
+  }, [determineLocation])
+
+  useEffect(() => {
+    if (useSettingsListener) {
+      const onLocationChanged = (enabled: boolean) => {
+        if (enabled) {
+          requestAndDetermineLocation()
+        } else {
+          setLocation(null)
+          setLocationState({
+            status: 'unavailable',
+            message: 'notAvailable'
+          })
+        }
+      }
+      const listener = SystemSetting.addLocationListener(onLocationChanged)
+      return () => {
+        listener.then(listener => listener && SystemSetting.removeListener(listener))
+      }
```

Comment:
This updates the location state, whenever the user changes the location permission.

Reply:
Perhaps you could add a comment :)